scio-avro/src/main/scala/com/spotify/scio/coders/avro/AvroCoders.scala

```diff
@@ -22,35 +22,47 @@ import com.spotify.scio.coders.{Coder, CoderGrammar}
 import com.spotify.scio.util.ScioUtil
 import org.apache.avro.Schema
 import org.apache.avro.generic.{GenericRecord, IndexedRecord}
+import org.apache.avro.io.{BinaryDecoder, BinaryEncoder, DecoderFactory, EncoderFactory}
 import org.apache.avro.specific.{SpecificData, SpecificFixed, SpecificRecord}
 import org.apache.beam.sdk.coders.Coder.NonDeterministicException
-import org.apache.beam.sdk.coders.{AtomicCoder, Coder => BCoder, CustomCoder, StringUtf8Coder}
+import org.apache.beam.sdk.coders.{CustomCoder, StringUtf8Coder}
 import org.apache.beam.sdk.extensions.avro.coders.AvroCoder
 import org.apache.beam.sdk.extensions.avro.io.AvroDatumFactory
-import org.apache.beam.sdk.util.common.ElementByteSizeObserver
+import org.apache.beam.sdk.util.EmptyOnDeserializationThreadLocal
 
 import java.io.{InputStream, OutputStream}
+import scala.collection.concurrent.TrieMap
 import scala.reflect.ClassTag
 
-final private class SlowGenericRecordCoder extends AtomicCoder[GenericRecord] {
-  // TODO: can we find something more efficient than String ?
+final private class SlowGenericRecordCoder extends CustomCoder[GenericRecord] {
+  // Schema is serializable on avro 1.9+
+  // Use String for 1.8 compat
   private val sc = StringUtf8Coder.of()
 
-  private def genericCoder(schema: Schema): BCoder[GenericRecord] =
-    AvroCoder.of(GenericRecordDatumFactory, schema)
+  // Reuse parsed schemas because avro uses caching
```

Comment:
nit: can you expand this comment a bit to explain what specifically is cached in Avro? I don't want us to forget the context on this later 🙏